src/backend/distributed/master/master_metadata_utility.c

```diff
@@ -62,6 +64,242 @@ static ShardPlacement * TupleToShardPlacement(TupleDesc tupleDesc,
+ * different size functions supported by PG are also supported by this function changing
+ * the size query given as the last parameter to function.
+ */
+StringInfo
+GenerateSizeQueryOnMultiplePlacements(Oid distributedRelationId, List *shardIntervalList,
+									  char *sizeQuery)
+{
+	/* build shard qualified name */
+	Oid schemaId = get_rel_namespace(distributedRelationId);
+	char *schemaName = get_namespace_name(schemaId);
+
+	StringInfo selectQuery = makeStringInfo();
+	bool isFirstShard = true;
+
+	ListCell *shardCell = NULL;
+
+	foreach(shardCell, shardIntervalList)
+	{
+		ShardInterval *shardInterval = (ShardInterval *) lfirst(shardCell);
+		uint64 shardId = shardInterval->shardId;
+		char *shardName = get_rel_name(distributedRelationId);
+		char *shardQualifiedName = NULL;
+		char *quotedShardName = NULL;
+		AppendShardIdToName(&shardName, shardId);
+
+		shardQualifiedName = quote_qualified_identifier(schemaName, shardName);
+		quotedShardName = quote_literal_cstr(shardQualifiedName);
+
+		if (isFirstShard)
+		{
+			appendStringInfo(selectQuery, "SELECT ");
+			isFirstShard = !isFirstShard;
+		}
+
+		appendStringInfo(selectQuery, sizeQuery, quotedShardName);
+	}
+
+	/* there is no more size function so just add 0 instead of it */
+	appendStringInfo(selectQuery, "0;");
+	return selectQuery;
+}
+
+
+/*
+ * SupportFunctionForCitusSize is a helper function for all citus size functions. It
+ * first checks whether the table is distributed and size query can be run on it.
+ * Connection to each node has to be established to get the size of the table.
+ */
+uint64
+SupportFunctionForCitusSize(Oid relationId, char *sizeQuery)
+{
+	MultiConnection *connection = NULL;
+	List *workerNodeList = WorkerNodeList();
+	ListCell *workerNodeCell = NULL;
+	uint64 totalRelationSize = 0;
+
+	/* check whether the size of table can be obtained with citus size functions */
+	ErrorIfNotSuitableToGetSize(relationId);
+
+	foreach(workerNodeCell, workerNodeList)
+	{
+		PGresult *queryResult = NULL;
+		StringInfo tableSizeQuery = NULL;
+		WorkerNode *workerNode = (WorkerNode *) lfirst(workerNodeCell);
+		char *workerNodeName = workerNode->workerName;
+		uint32 workerNodePort = workerNode->workerPort;
+		char *tableSizeString;
+		int executeCommand = 0;
+		uint32 connectionFlag = 0;
+
+		List *intervalsOnNode = GetShardIntervalsOnNode(workerNode, relationId);
+
+
+		tableSizeQuery = GenerateSizeQueryOnMultiplePlacements(relationId,
+															   intervalsOnNode,
+															   sizeQuery);
+
+		connection = GetNodeConnection(connectionFlag, workerNodeName, workerNodePort);
+
+		executeCommand = ExecuteOptionalRemoteCommand(connection, tableSizeQuery->data,
+													  &queryResult);
+
+		/*
+		 * If there is a connection error and the table is not reference table,
+		 * close the connection and quit. If it is a reference table continue
+		 * to get the size from another node.
+		 */
+		if (executeCommand != 0)
+		{
+			PQclear(queryResult);
```

Comment:
When you error out, forgetting results etc. automatically handled. You do not need to call PQclear and ForgetResults.